### PR TITLE
Oauth DCR

### PR DESCRIPTION
**Pull request by harsha1979**

### CODE REVIEW

File: features/org.wso2.carbon.identity.oauth.dcr.server.feature/pom.xml [hunk 1]
```diff
@@ -0,0 +1,183 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (thanujalk) [inline features/org.wso2.carbon.identity.oauth.dcr.server.feature/pom.xml]: change year to 2016

File: components/org.wso2.carbon.identity.oauth.dcr/pom.xml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <!--dependency>
+            <groupId>javax.ws.rs</groupId>
+            <artifactId>jsr311-api</artifactId>
+            <scope>provided</scope>
+        </dependency-->
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.dcr/pom.xml]: remove this commented dependency.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/internal/DCRServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                                                                new HttpUnregistrationResponseFactory(), null);
+
+            componentContext.getBundleContext().registerService(RegistrationHandler.class.getName(),
+                                                                new RegistrationHandler(), null);
+
+            componentContext.getBundleContext().registerService(UnRegistrationHandler.class.getName(),
+                                                                new UnRegistrationHandler(), null);
+
+
+            System.out.print(">>>>>");
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/internal/DCRServiceComponent.java]: remove this line.

File: components/org.wso2.carbon.identity.oidc.dcr/src/main/java/org/wso2/carbon/identity/oidc/dcr/internal/OIDCDCRServiceComponent.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+public class OIDCDCRServiceComponent {
+
+    private static final Log log = LogFactory.getLog(OIDCDCRServiceComponent.class);
+
+    @SuppressWarnings("unused")
+    protected void activate(ComponentContext componentContext) {
+        try {
+            System.out.print("DDD");
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oidc.dcr/src/main/java/org/wso2/carbon/identity/oidc/dcr/internal/OIDCDCRServiceComponent.java]: remove this line

File: components/org.wso2.carbon.identity.oauth.dcr/pom.xml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                            org.wso2.carbon.identity.application.common.model,
+                            org.wso2.carbon.identity.application.mgt,
+                            org.wso2.carbon.identity.application.mgt.stub,
+                            org.wso2.carbon.identity.base,
+                            org.wso2.carbon.identity.sso.saml.admin,
+                            org.wso2.carbon.identity.sso.saml.dto,
+                            org.wso2.carbon.registry.api,
+                            org.wso2.carbon.registry.core,
+                            org.wso2.carbon.utils.multitenancy,
+                            org.wso2.carbon.identity.application.authentication.framework.*,
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.dcr/pom.xml]: Specify import version ranges.

File: components/org.wso2.carbon.identity.oidc.dcr/pom.xml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                            org.wso2.carbon.identity.application.mgt,
+                            org.wso2.carbon.identity.application.mgt.stub,
+                            org.wso2.carbon.identity.base,
+                            org.wso2.carbon.identity.sso.saml.admin,
+                            org.wso2.carbon.identity.sso.saml.dto,
+                            org.wso2.carbon.registry.api,
+                            org.wso2.carbon.registry.core,
+                            org.wso2.carbon.utils.multitenancy,
+                            org.wso2.carbon.identity.oauth.dcr,
+                            org.wso2.carbon.identity.application.authentication.framework.*,
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oidc.dcr/pom.xml]: Specify import version ranges

File: components/org.wso2.carbon.identity.oauth.dcr/pom.xml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.json.wso2</groupId>
+            <artifactId>json</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.googlecode.json-simple.wso2</groupId>
+            <artifactId>json-simple</artifactId>
+            <version>1.1.wso2v1</version>
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/pom.xml]: Specify dependency version with a property

File: components/org.wso2.carbon.identity.oidc.dcr/pom.xml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.json.wso2</groupId>
+            <artifactId>json</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.googlecode.json-simple.wso2</groupId>
+            <artifactId>json-simple</artifactId>
+            <version>1.1.wso2v1</version>
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oidc.dcr/pom.xml]: add property for version

File: components/org.wso2.carbon.identity.oidc.dcr/src/main/java/org/wso2/carbon/identity/oidc/dcr/internal/OIDCDCRServiceComponent.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            componentContext.getBundleContext().registerService(HttpIdentityRequestFactory.class.getName(),
+                                                                new OIDCRegistrationRequestFactory(), null);
+            componentContext.getBundleContext().registerService(IdentityProcessor.class.getName(),
+                                                                new OIDCDCRProcessor(), null);
+            componentContext.getBundleContext().registerService(HttpIdentityResponseFactory.class.getName(),
+                                                                new HttpOIDCRegistrationResponseFactory(), null);
+
+
+        } catch (Exception ee) {
+            ee.printStackTrace();
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oidc.dcr/src/main/java/org/wso2/carbon/identity/oidc/dcr/internal/OIDCDCRServiceComponent.java]: log the error

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/internal/DCRServiceComponent.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        } catch (Exception ee) {
+            ee.printStackTrace();
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/internal/DCRServiceComponent.java]: log the error.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRClientException.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.dcr;
+
+
+import org.wso2.carbon.identity.application.authentication.framework.inbound.FrameworkClientException;
+
+public class DCRClientException extends FrameworkClientException {
+    protected DCRClientException(String errorDescription) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRClientException.java]: why is this protected?

Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRClientException.java]: why is this protected?

Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRClientException.java]: why is this protected?

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/HttpRegistrationResponseFactory.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+import org.wso2.carbon.identity.oauth.common.OAuthConstants;
+import org.wso2.carbon.identity.oauth.dcr.exception.RegistrationException;
+import org.wso2.carbon.identity.oauth.dcr.model.RegistrationResponse;
+import org.wso2.carbon.identity.oauth.dcr.util.ErrorCodes;
+
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import java.util.List;
+
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/HttpRegistrationResponseFactory.java]: add class level comment

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/HttpUnregistrationResponseFactory.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import org.wso2.carbon.identity.application.authentication.framework.inbound.HttpIdentityResponse;
+import org.wso2.carbon.identity.application.authentication.framework.inbound.HttpIdentityResponseFactory;
+import org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityResponse;
+import org.wso2.carbon.identity.oauth.common.OAuthConstants;
+import org.wso2.carbon.identity.oauth.dcr.exception.UnRegistrationException;
+import org.wso2.carbon.identity.oauth.dcr.model.UnregistrationResponse;
+
+import javax.servlet.http.HttpServletResponse;
+
+
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/HttpUnregistrationResponseFactory.java]: add class level comment.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/UnregistrationRequestFactory.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+import org.wso2.carbon.identity.oauth.dcr.util.DCRConstants;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.HttpHeaders;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/UnregistrationRequestFactory.java]: add class level comment.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRManagementService.java [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(userName);
+
+        OAuthAdminService oAuthAdminService = new OAuthAdminService();
+        OAuthConsumerAppDTO oAuthConsumerApp = null;
+        try {
+            oAuthConsumerApp = oAuthAdminService.getOAuthApplicationData(consumerKey);
+        } catch (Exception e) {
+            //We had to catch Exception here because getOAuthApplicationData can throw exceptions of java.lang.Exception
+            // class.
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRManagementService.java]: add a debug/error log

File: features/org.wso2.carbon.identity.oauth.dcr.server.feature/pom.xml [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+                            <id>org.wso2.carbon.identity.oauth.dcr.server</id>
+                            <propertiesFile>../etc/feature.properties</propertiesFile>
+                            <adviceFile>
+                                <properties>
+                                    <propertyDef>org.wso2.carbon.p2.category.type:server</propertyDef>
+                                    <propertyDef>org.eclipse.equinox.p2.type.group:false</propertyDef>
+                                </properties>
+                            </adviceFile>
+                            <bundles>
+                                <!--bundleDef>com.google.gdata.wso2:gdata-core</bundleDef>
```
Comment (mefarazath) [inline features/org.wso2.carbon.identity.oauth.dcr.server.feature/pom.xml]: remove commented bundleDefs.

**Comment by thanujalk** [comment]
Merging without fixing code review suggestions. Will provide a PR with fixes.

**Comment by thanujalk** [comment]
Merging without fixing code review suggestions. Will provide a PR with fixes.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRClientException.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.dcr;
+
+
+import org.wso2.carbon.identity.application.authentication.framework.inbound.FrameworkClientException;
+
+public class DCRClientException extends FrameworkClientException {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRClientException.java]: Missing class level comment.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRRuntimeException.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.dcr;
+
+
+import org.wso2.carbon.identity.base.IdentityRuntimeException;
+
+public class DCRRuntimeException extends IdentityRuntimeException {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/DCRRuntimeException.java]: Missing class level comment.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/context/DCRMessageContext.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package org.wso2.carbon.identity.oauth.dcr.context;
+
+import org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityRequest;
+import org.wso2.carbon.identity.core.bean.context.MessageContext;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DCRMessageContext extends MessageContext {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/context/DCRMessageContext.java]: Missing class level comment.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/exception/RegistrationException.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
+package org.wso2.carbon.identity.oauth.dcr.exception;
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/exception/RegistrationException.java]: Missing licence header.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/exception/RegistrationException.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+
+
+import org.wso2.carbon.identity.oauth.dcr.DCRException;
+
+public class RegistrationException extends DCRException {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/exception/RegistrationException.java]: Missing class level comment.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/exception/UnRegistrationException.java [hunk 22]
```diff
@@ -0,0 +1,26 @@
+package org.wso2.carbon.identity.oauth.dcr.exception;
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/exception/UnRegistrationException.java]: Missing licence header.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/util/DCRExceptionBuilder.java [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public static <T extends IdentityException> T buildException(Class<T> exceptionClass, String errorCode,
+                                                                 String errorDescription) throws DCRRuntimeException {
+        try {
+            T exception = exceptionClass.getConstructor(String.class).newInstance(errorDescription);
+            return buildException(exception, errorCode, errorDescription);
+        } catch (Exception e) {
+            throw new DCRRuntimeException(
+                    "Error occurred while creating new instant using reflection, " + e.getMessage());
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/util/DCRExceptionBuilder.java]: Should be 'instance' ?

**Pull request merged**